### PR TITLE
[api] support rebuild and release triggers

### PR DESCRIPTION
**Pull request by adrianschroeter**
required for CR releases by automated testing and for openSUSE ports
co-maintainers

### CODE REVIEW

File: src/api/app/controllers/trigger_controller.rb [hunk 1]
```diff
@@ -1,16 +1,71 @@
...
   skip_before_action :require_login
 
-  # github.com sends a hash payload
-  skip_before_action :validate_params, only: [:runservice]
+  def rebuild
+    Backend::Api::Sources::Package.rebuild(@pkg.project.name, @pkg.name)
+    render_ok
+  end
+
+  # rubocop:disable Lint/NonLocalExitFromIterator
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: why? :sob:

File: src/api/app/controllers/trigger_controller.rb [hunk 2]
```diff
@@ -1,16 +1,71 @@
...
+      render_error message: 'Token must define the release package',
+                   status: 403,
+                   errorcode: 'no_permission'
+      return
+    end
+
+    matched_repo = false
+    User.current = @token.user
+    @pkg.project.repositories.each do |repo|
+      repo.release_targets.each do |releasetarget|
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: N+1 queries! :see_no_evil: (performance improvement)

File: src/api/app/controllers/trigger_controller.rb [hunk 3]
```diff
@@ -1,16 +1,71 @@
...
+        next unless releasetarget.trigger == 'manual'
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: You can add this to the SQL query :wink: (performance improvement)

File: src/api/app/controllers/trigger_controller.rb [hunk 4]
```diff
@@ -1,16 +1,71 @@
...
+        unless @token.user.can_modify?(releasetarget.repository.project)
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: `repo.project`

You are using the project several time (some of them as `@pkg.project`). I think it is a good idea to assign the project to a variable before the loop :wink:

File: src/api/app/controllers/trigger_controller.rb [hunk 5]
```diff
@@ -1,16 +1,71 @@
...
+          render_error message: "no permission for target #{releasetarget.repository.project.name}",
+                       status: 403,
+                       errorcode: 'no_permission'
+          return
+        end
+        target_package_name = @pkg.release_target_name
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: This is the same every time you iterate, you should put it outside the iteration (small performance improvement)

File: src/api/app/controllers/trigger_controller.rb [hunk 6]
```diff
@@ -1,16 +1,71 @@
...
+
+        # find md5sum and release source and binaries
+        release_package(@pkg, releasetarget.target_repository, target_package_name, repo, nil, nil, nil, true)
+        matched_repo = true
+      end
+    end
+    unless matched_repo
+      render_error message: "no repository from #{@pkg.project.name} could get released",
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: [nitpick] `.name` is not needed

File: src/api/app/controllers/person/token_controller.rb [hunk 7]
```diff
@@ -18,14 +18,16 @@ def create
       if params[:project] || params[:package]
         pkg = Package.get_by_project_and_name(params[:project], params[:package])
       end
-      @token = Token::Service.create(user: @user, package: pkg)
+      @token = Token::Service.create(user: @user, package: pkg) if params[:operation].blank? || params[:operation] == 'runservice'
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/controllers/person/token_controller.rb]: [nitpick] And it would be a good idea to move the `token_type` part to the token model as it is also used in https://github.com/openSUSE/open-build-service/pull/6965/files#diff-00232d8e1ee1f7438e6cbb963c74a038R82 and then it could be reused :smile:

File: src/api/app/controllers/trigger_controller.rb [hunk 8]
```diff
@@ -19,35 +74,32 @@ def runservice
...
     unless auth && auth[0..4] == 'Token' && auth[6..-1] =~ /^[A-Za-z0-9+\/]+$/
       render_error errorcode: 'permission_denied',
-                   message: "No valid token found 'Authorization' header",
+                   message: 'No valid token found "Authorization" header',
                    status: 403
       return
     end
 
-    token = Token::Service.find_by_string(auth[6..-1])
+    token_type = Token::Service
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: [nitpick] this is exactly was `case` is for :wink: 

```ruby
token_type = case request['action']
             when 'rebuild'
               Token::Rebuild
             when 'release'
               Token::Release
             else
               Token::Service
             end

File: src/api/app/controllers/trigger_controller.rb [hunk 9]
```diff
@@ -19,35 +74,32 @@ def runservice
...
+    return unless @pkg
 
-    # execute the service in backend
-    path = pkg.source_path
-    params = { cmd: 'runservice', comment: 'runservice via trigger', user: token.user.login }
-    path << build_query_from_hash(params, [:cmd, :comment, :user])
-    pass_to_backend(path)
+    # check if user has still access
+    unless @token.user.is_active? && @token.user.can_modify?(@pkg)
+      render_error message: "no permission for package #{@pkg.name} in project #{@pkg.project.name}",
```
Comment (Ana06) [inline src/api/app/controllers/trigger_controller.rb]: [nitpick] `.name` is not needed

File: src/api/app/models/token.rb [hunk 10]
```diff
@@ -1,10 +1,14 @@
 class Token < ApplicationRecord
-  belongs_to :user, foreign_key: 'user_id', inverse_of: :service_tokens
+  belongs_to :user, foreign_key: 'user_id'
```
Comment (Ana06) [inline src/api/app/models/token.rb]: why is `foreign_key` needed? :thinking:

File: src/api/app/models/token.rb [hunk 11]
```diff
@@ -1,10 +1,14 @@
...
   belongs_to :package, inverse_of: :tokens
 
   has_secure_token :string
 
   validates :user, presence: true
+
+  def token_name
```
Comment (Ana06) [inline src/api/app/models/token.rb]: why don't you just define it as an instance method in the subclass? :confused:

File: src/api/app/controllers/trigger_controller.rb (hunk 1 above)
Comment (adrianschroeter) [reply src/api/app/controllers/trigger_controller.rb]: because we must escape from the loops with proper error message and don't want to have additional checking code just to get away with break.

File: src/api/app/controllers/trigger_controller.rb (hunk 2 above)
Comment (adrianschroeter) [reply src/api/app/controllers/trigger_controller.rb]: this is really absolute minor here ... these are release jobs, they are expensive....

File: src/api/app/controllers/trigger_controller.rb (hunk 1 above)
Comment (Ana06) [reply src/api/app/controllers/trigger_controller.rb]: ok

File: src/api/app/controllers/trigger_controller.rb (hunk 2 above)
Comment (Ana06) [reply src/api/app/controllers/trigger_controller.rb]: well, this is raising a lot of extra queries and it is easy to solve:

```ruby
@pkg.project.repositories.includes(:release_targets).each do |repo|

File: src/api/app/models/token.rb (hunk 11 above)
Comment (Ana06) [reply src/api/app/models/token.rb]: I am not sure if it clear what I meant. You could define for example in `rebuild.rb`:

```
def token_name
  'rebuild'
end
```

without the `self`

**Comment by hennevogel** [comment]
@adrianschroeter can we help you to push this over the line somehow?

File: src/api/app/models/token.rb (hunk 11 above)
Comment (Ana06) [reply src/api/app/models/token.rb]: ^^

File: src/api/app/controllers/trigger_controller.rb (hunk 5 above)
Comment (Ana06) [reply src/api/app/controllers/trigger_controller.rb]: ^^

Comment (Ana06) [reply src/api/app/controllers/trigger_controller.rb]: This is the same every time you iterate, you should put it outside the iteration (small performance improvement)

File: src/api/app/models/token.rb (hunk 11 above)
Comment (Ana06) [reply src/api/app/models/token.rb]: his could be an instance method of the subclass. Any reason to do it like this needing to do `self.class.token_name`?

File: src/api/app/controllers/person/token_controller.rb [hunk 12]
```diff
@@ -7,25 +7,34 @@ class TokenController < ApplicationController
...
+
+      token_type = case params[:operation]
+                   when 'rebuild'
+                     Token::Rebuild
+                   when 'release'
+                     Token::Release
+                   else
+                     # default is Token::Service
+                     Token::Service
+                   end
```
Comment (bgeuken) [inline src/api/app/controllers/person/token_controller.rb]: This looks like it could be dried with the code [here](https://github.com/openSUSE/open-build-service/pull/6965/files#diff-00232d8e1ee1f7438e6cbb963c74a038R83)

Comment (vpereira) [reply src/api/app/controllers/person/token_controller.rb]: yup. Did it now.

File: src/api/app/controllers/person/token_controller.rb [hunk 13]
```diff
@@ -7,25 +7,25 @@ class TokenController < ApplicationController
...
     # POST /person/<login>/token
     def create
       authorize @user, :update?
 
-      pkg = nil
-      if params[:project] || params[:package]
-        pkg = Package.get_by_project_and_name(params[:project], params[:package])
-      end
-      @token = Token::Service.create(user: @user, package: pkg)
+      pkg = if params[:project] || params[:package]
```
Comment (Ana06) [inline src/api/app/controllers/person/token_controller.rb]: [nitpick] this refactoring is unrelated

File: src/api/app/controllers/trigger_controller.rb (hunk 5 above)
Comment (Ana06) [reply src/api/app/controllers/trigger_controller.rb]: this haven't been addressed

File: src/api/app/models/token.rb (hunk 11 above)
Comment (Ana06) [reply src/api/app/models/token.rb]: this  is also not addressed...

Comment (bgeuken) [reply src/api/app/models/token.rb]: IIRC, this could also be a delegate to the class, but I think that's more of a nitpick

File: src/api/app/models/token.rb [hunk 14]
```diff
@@ -1,10 +1,26 @@
...
+    self.class.token_name
+  end
+
+  def self.token_type(action)
```
Comment (Ana06) [inline src/api/app/models/token.rb]: this can be returned directly in the methods.

File: src/api/app/controllers/trigger_controller.rb [hunk 15]
```diff
@@ -1,53 +1,96 @@
...
 
-  def runservice
-    auth = request.env['HTTP_AUTHORIZATION']
+  # rubocop:disable Lint/NonLocalExitFromIterator
+  def release
+    matched_repo = false
+    User.session = @token.user
+    @pkg.project.repositories.includes(:release_targets).each do |repo|
+      repo.release_targets.where(trigger: 'manual').each do |releasetarget|
+        unless @token.user.can_modify?(releasetarget.repository.project)
```
Comment (bgeuken) [inline src/api/app/controllers/trigger_controller.rb]: If I am not mistaken, we want to check the `target_repository` of the release target instead: `releasetarget.target_repository.project`

Comment (adrianschroeter) [reply src/api/app/controllers/trigger_controller.rb]: right, this is indeed wrong

File: src/api/app/controllers/trigger_controller.rb [hunk 16]
```diff
@@ -1,53 +1,96 @@
...
+    Backend::Api::Sources::Package.rebuild(@pkg.project.name, @pkg.name)
+    render_ok
+  end
 
-  def runservice
-    auth = request.env['HTTP_AUTHORIZATION']
+  # rubocop:disable Lint/NonLocalExitFromIterator
+  def release
+    matched_repo = false
+    User.session = @token.user
```
Comment (adrianschroeter) [inline src/api/app/controllers/trigger_controller.rb]: IMHO we should not do this, because any programming error or refactoring may lead to a security issue.
The token should only allow a certain operation, but we should prepare for side effects.

The problem coolo reported is actually a generic problem that we always try to write the package meta when using Package.source_changed. source_changed was supposed just to update database content which could get replayed from the source server.
However, this is beyond the topic of this pull request and a generic problem atm.

File: src/api/app/controllers/trigger_controller.rb (hunk 15 above)
Comment (vpereira) [reply src/api/app/controllers/trigger_controller.rb]: @adrian 

so i have to check the permission on `releasetarget.target_repository.project` ?

what else is wrong? the call to `release_package` is right or there is something wrong there too?

File: src/api/app/controllers/trigger_controller.rb (hunk 16 above)
Comment (vpereira) [reply src/api/app/controllers/trigger_controller.rb]: I think the issue isn't about the controller, but operations on models and backend that need an user.. I think thats horrible too, but I'm not sure we are able to avoid it now.

Comment (coolo) [reply src/api/app/controllers/trigger_controller.rb]: create a backlog item and continue :(

**Comment by adrianschroeter** [comment]
On Freitag, 17. Mai 2019, 17:00:51 CEST Victor Pereira wrote:
> vpereira commented on this pull request.
> 
> 
> 
> 
> -  def runservice
> -    auth = request.env['HTTP_AUTHORIZATION']
> +  # rubocop:disable Lint/NonLocalExitFromIterator
> +  def release
> +    matched_repo = false
> +    User.session = @token.user
> +    @pkg.project.repositories.includes(:release_targets).each do |repo|
> +      repo.release_targets.where(trigger: 'manual').each do |releasetarget|
> +        unless @token.user.can_modify?(releasetarget.repository.project)
> 
> @adrian
> 
> so i have to check the permission on
> `releasetarget.target_repository.project` ?

yes! Just replace that line to

  unless @token.user.can_modify?(releasetarget.target_repository.project)

> what else is wrong? the call to `release_package` is right or there is
> something wrong there too?

should be fine and worked in the test suite stub from original
submission at least

-- 

Adrian Schroeter
SUSE Linux Products GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
email: adrian@suse.de

**Comment by adrianschroeter** [comment]
On Freitag, 17. Mai 2019, 17:02:02 CEST Victor Pereira wrote:
> vpereira commented on this pull request.
> 
> 
> 
> 
> -  def runservice
> -    auth = request.env['HTTP_AUTHORIZATION']
> +  # rubocop:disable Lint/NonLocalExitFromIterator
> +  def release
> +    matched_repo = false
> +    User.session = @token.user
> 
> I think the issue isn't about the controller, but operations on models and
> backend that need an user.. I think thats horrible too, but I'm not sure we
> are able to avoid it now.

I agree that this is an independend problem. This pull request is not
changing anything here.


-- 

Adrian Schroeter
SUSE Linux Products GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
email: adrian@suse.de

File: src/api/app/controllers/trigger_controller.rb (hunk 2 above)
Comment (vpereira) [reply src/api/app/controllers/trigger_controller.rb]: this is resolved, right?

File: src/api/app/models/token.rb (hunk 11 above)
Comment (vpereira) [reply src/api/app/models/token.rb]: we didn't want to change it before we have specs for that.

File: src/api/app/controllers/person/token_controller.rb (hunk 13 above)
Comment (vpereira) [reply src/api/app/controllers/person/token_controller.rb]: yes, since we were refactoring other things, this nitpick isn't really relevant.

File: src/api/app/controllers/trigger_controller.rb (hunk 5 above)
Comment (vpereira) [reply src/api/app/controllers/trigger_controller.rb]: AFAIR in your original change request, you handled it as nitpick. So, because Adrian didn't want to change it, I will keep it too (probably because the number of release_targets where trigger = manual won't end up in thousand lines iteration)

File: src/api/app/models/token.rb (hunk 11 above)
Comment (vpereira) [reply src/api/app/models/token.rb]: I'm not sure what would be the difference here, but because it was something that Adrian did, and I don't see why we should change it now, I would keep it (unless there is a strong argument to change it)

File: src/api/app/models/token.rb (hunk 14 above)
Comment (vpereira) [reply src/api/app/models/token.rb]: @Ana06 could be more specific, please?

File: src/api/app/controllers/trigger_controller.rb (hunk 4 above)
Comment (vpereira) [reply src/api/app/controllers/trigger_controller.rb]: because it's a nitpick, I would let as it is, since the number of matches is really specific, I would keep it as @adrianschroeter did it and optimize it if necessary later... what do you think?

**Comment by vpereira** [comment]
> There are still not two addressed changes. Apart from that LGTM

The two not addressed change requests are, as you wrote, nitpick, and I wouldn't change it now unless there is a strong reason to do it.

**Comment by vpereira** [comment]
> I wrote two comments that are not nitpick:
> 
> [#6965 (comment)](https://github.com/openSUSE/open-build-service/pull/6965#discussion_r283311990)
> 
> [#6965 (comment)](https://github.com/openSUSE/open-build-service/pull/6965#discussion_r256834313)
> 
> But if it is needed to merge this, go ahead. I can send a PR afterwards to improve it.

great, if you want later when you are refactoring it, you can ping me and we can do it together. Thank you @Ana06 !

File: src/api/app/controllers/trigger_controller.rb [hunk 17]
```diff
@@ -1,53 +1,85 @@
...
+    Backend::Api::Sources::Package.rebuild(@pkg.project.name, @pkg.name)
+    render_ok
+  end
 
-    token = Token::Service.find_by_string(auth[6..-1])
+  def release
+    matched_repo = false
+    @pkg.project.repositories.includes(:release_targets).each do |repo|
+      repo.release_targets.where(trigger: 'manual').each do |releasetarget|
+        release_target_repository_project = releasetarget.repository.project
```
Comment (bgeuken) [inline src/api/app/controllers/trigger_controller.rb]: This should be `releasetarget.target_repository.project`. releasetarget.repository would be the `repo` we are coming from

Comment (vpereira) [reply src/api/app/controllers/trigger_controller.rb]: could please take a look again @bgeuken ?
